README.md
<p>
    <b>This is migrated to <a href="https://github.com/javaxcel/javaxcel">https://github.com/javaxcel/javaxcel</a>.</b>
</p>

<hr>

<p align="center">
    <img src="./asset/javaxcel-core-logo.png" alt="Javaxcel Core" width="20%">
</p>

<h1 align="center">Javaxcel Core</h1>

<p align="center">Supporter for export and import of excel file</p>

<p align="center">
    <!--
    <a href="https://travis-ci.com/github/javaxcel/javaxcel-core">
        <img alt="Travis CI" src="https://img.shields.io/travis/com/javaxcel/javaxcel-core/release?style=flat-square"/>
    </a>
    -->
    <a href="https://github.com/javaxcel/javaxcel-core/actions/workflows/maven-build.yml">
        <img alt="Github Workflow" src="https://img.shields.io/github/workflow/status/javaxcel/javaxcel-core/Java%20CI%20with%20Maven/release?style=flat-square">
    </a>
    <a href="https://codecov.io/gh/javaxcel/javaxcel-core">
        <img alt="Codecov branch" src="https://img.shields.io/codecov/c/github/javaxcel/javaxcel-core/release?label=code%20coverage&style=flat-square&token=X7ZO535W9K"/>
    </a>
    <a href="https://search.maven.org/artifact/com.github.javaxcel/javaxcel-core">
        <img alt="Maven Central" src="https://img.shields.io/maven-central/v/com.github.javaxcel/javaxcel-core?style=flat-square">
    </a>
    <br/>
    <a href="https://lgtm.com/projects/g/javaxcel/javaxcel-core/context:java">
        <img alt="Lgtm grade" src="https://img.shields.io/lgtm/grade/java/github/javaxcel/javaxcel-core.svg?logo=&logoWidth=&label=lgtm%3A%20code%20quality&&style=flat-square"/>
    </a>
    <a href="https://www.codacy.com/gh/javaxcel/javaxcel-core/dashboard">
        <img alt="Codacy grade" src="https://img.shields.io/codacy/grade/a430e1580c5f44d2b2f1b2c11195cc49?label=codacy%3A%20code%20quality&style=flat-square">
    </a>
    <img alt="jdk8" src="https://img.shields.io/badge/jdk-8-orange?style=flat-square">
</p>

## Table of Contents

- [What is Javaxcel?](#what-is-javaxcel)
- [Getting started](#getting-started)
- [Examples](#examples)
    1. [No option](#no-option)
    2. [Exclude field](#exclude-field)
    3. [Name the header](#name-the-header)
    4. [Set the default value](#set-the-default-value)
    5. [Model without the targeted fields](#model-without-the-targeted-fields)
    6. [Model that extends class](#model-that-extends-class)
    7. [Format date/time](#format-datetime)
    8. [Name a Sheet](#name-a-sheet)
    9. [Decoration](#decoration)
    10. [Expression](#expression)
    11. [Enum value constraint](#enum-value-constraint)
    12. [Limitation of reading rows](#limitation-of-reading-rows)
    13. [Parallel reading](#parallel-reading)
    13. [Resolution of constructor and method](#resolution-of-constructor-and-method)
    13. [Add a handler for custom type](#add-a-handler-for-custom-type)
    14. [Support java.util.Map](#support-javautilmap)

<br><br>

# What is Javaxcel?

Javaxcel is a supporter for exporting `java.util.List` to spreadsheets and importing `java.util.List` from spreadsheets using [Apache POI](https://github.com/apache/poi).

<br><br>

# Getting started

### Maven

```xml
<dependency>
    <groupId>com.github.javaxcel</groupId>
    <artifactId>javaxcel-core</artifactId>
    <version>x.y.z</version>
</dependency>

<!-- Required dependency -->
<dependency>
    <groupId>org.apache.poi</groupId>
    <artifactId>poi-ooxml</artifactId>
    <version>a.b.c</version>
</dependency>
```

### Gradle

```groovy
implementation 'com.github.javaxcel:javaxcel-core:x.y.z'

// Required dependency
implementation 'org.apache.poi:poi-ooxml:a.b.c'
```

<br>

```java
// Creates an instance of Javaxcel.
Javaxcel javaxcel = Javaxcel.newInstance();

File src = new File("/data", "old-products.xls");
File dest = new File("/data", "new-products.xlsx");

try (InputStream in = new FileInputStream(src);
        OutputStream out = new FileOutputStream(dest);
        Workbook oldWorkbook = new HSSFWorkbook(in);
        Workbook newWorkbook = new SXSSFWorkbook()) {
    // Reads all the sheets and returns data as a list.
    List<Product> products = javaxcel.reader(oldWorkbook, Product.class).read();
    
    // Creates an Excel file and writes data to it.
    javaxcel.writer(newWorkbook, Product.class).write(out, products);
} catch (IOException e) {
    e.printStackTrace();
}
```

Code with simple usage.

<br><br>

# Examples

```java
class Product {
    private long serialNumber;
    private String name;
    private String accessId;
    private Double width;
    private double depth;
    private double height;
    private Double weight;
}

// ...

Product product = Product.builder()
        .serialNumber(10000)
        .name("Choco cereal")
        .accessId("2a60-4973-aec0-685e")
        .height(20.5)
        .weight(580.5)
        .build();
List<Product> products = Collections.singletonList(product);
```

There is a list that contains a `Product`.

<br><br>

## No option

### writer:

```java
File dest = new File("/data", "products.xlsx")
OutputStream out = new FileOutputStream(dest);
Workbook workbook = new SXSSFWorkbook();

Javaxcel.newInstance()
        .writer(workbook, Product.class)
        .write(out, products);
```

The result is

| serialNumber | name         | accessId            | width | depth | height | weight |
| ------------ | ------------ | ------------------- | ----- | ----- | ------ | ------ |
| 10000        | Choco cereal | 2a60-4973-aec0-685e |       | 0.0   | 20.5   | 580.5  |

The column order is the same as the declared field order.

If nothing is specified for the column, header name is the field name.

<br>

### reader:

```java
File src = new File("/data", "products.xlsx");
Workbook workbook = new XSSFWorkbook(src);

List<Product> products = Javaxcel.newInstance()
        .reader(workbook, Product.class)
        .read();
```

The result is

```json
[
  {
    "serialNumber": 10000,
    "name": "Choco cereal",
    "accessId": "2a60-4973-aec0-685e",
    "width": null,
    "depth": 0.0,
    "height": 20.5,
    "weight": 580.5
  }
]
```

The column order, also when read, is the same as the declared field order.

Model must has a constructor without parameters, so that `ExcelReader` can instantiate.

<br><br>

## Exclude field

```java
@ExcelIgnore
private String accessId;
```

### writer:

| serialNumber | name         | width | depth | height | weight |
| ------------ | ------------ | ----- | ----- | ------ | ------ |
| 10000        | Choco cereal |       | 0.0   | 20.5   | 580.5  |

If you want to exclude several fields, annotate `@ExcelIgnore` to them.

<br>

### reader:

```json
[
  {
    "serialNumber": 10000,
    "name": "Choco cereal",
    "accessId": null,
    "width": null,
    "depth": 0.0,
    "height": 20.5,
    "weight": 580.5
  }
]
```

`ExcelReader` will pass the fields that annotated `@ExcelIgnore` by.

If column `accessId` exists in Excel sheet and `Product#accessId` is still annotated `@ExcelIgnore`,

the exception will occur because of setting `accessId` to `width` (NumberFormatException).

<br><br>

## Name the header

```java
@ExcelColumn(name = "PRODUCT_NO")
private long serialNumber;

@ExcelColumn
private String name;
```

### writer:

| PRODUCT_NO | name         | accessId            | width | depth | height | weight |
| ---------- | ------------ | ------------------- | ----- | ----- | ------ | ------ |
| 10000      | Choco cereal | 2a60-4973-aec0-685e |       | 0.0   | 20.5   | 580.5  |

If you want to name the header, annotate `@ExcelColumn` and assign `name()` you want.

<br>

If you want to override `ExcelColumn#name()` (use header names on this moment),
invoke `ExcelWriter#options(ExcelWriteStrategy...)` with `ExcelWriteStrategy.HeaderNames`.

```java
List<String> headerNames = Arrays.asList("PRD_NO","NM","ACS_ID","WID","DEP","HEI","WEI"); // 7
// List<String> headerNames = Arrays.asList("PRD_NO","NM","ACS_ID","WID","DEP","HEI"); // 6: Occurs exception.

Javaxcel.newInstance()
        .writer(workbook, Product.class)
        .options(new HeaderNames(headerNames))
        .write(out, products);
```

The result is

| PRD_NO | NM           | ASC_ID              | WID  | DEP  | HEI  | WEI   |
| ------ | ------------ | ------------------- | ---- | ---- | ---- | ----- |
| 10000  | Choco cereal | 2a60-4973-aec0-685e |      | 0.0  | 20.5 | 580.5 |

If the number of arguments is not equal to the number of targeted fields, `ExcelWriter` throws exception.

<br><br>

## Set the default value

```java
// Default value is effective except primitive type.
@ExcelColumn(name = "WIDTH", defaultValue = "0.0mm")
private Double width;

// Default value is ineffective to primitive type.
@ExcelColumn(name = "Depth", defaultValue = "(empty)")
private double depth;

@ExcelColumn(defaultValue = "0")
private double height;
```

### writer:

| serialNumber | name         | accessId            | WIDTH | Depth | height | weight |
| ------------ | ------------ | ------------------- | ----- | ----- | ------ | ------ |
| 10000        | Choco cereal | 2a60-4973-aec0-685e | 0.0mm | 0.0   | 20.5   | 580.5  |

It's ineffective to assign default value to primitive type, because the field of primitive type is always initialized.

<br>

If you want to override `ExcelColumn#defaultValue()` (use another default value on this moment),
invoke `ExcelWriter#options(ExcelWriteStrategy...)` with option `ExcelWriteStrategy.DefaultValue`.

```java
Product product = new Product(); // Empty product.
List<Product> products = Collections.singletonList(product);

Javaxcel.newInstance()
        .writer(workbook, Product.class)
        .options(new DefaultValue("(empty)"))
        .write(out, products);
```

The result is

| serialNumber | name    | accessId | WIDTH   | Depth | height | weight  |
| ------------ | ------- | -------- | ------- | ----- | ------ | ------- |
| 0            | (empty) | (empty)  | (empty) | 0.0   | 0.0    | (empty) |

`ExcelWriteStrategy.DefaultValue(String)` will be applied to all fields.

<br>

### reader:

If cell value is null or empty string, `ExcelReadHandlerConverter` parse `ExcelColumn#defaultValue()` and set it to field.

Specify `ExcelColumn#defaultValue()` with care.
It might not occur exception, when you write, though the default value doesn't match type of that field.
It must occur exception, when you read, if the default value doesn't.

If the field that is annotated by `@ExcelReadExpression`, its default value doesn't work.

<br><br>

## Model without the targeted fields

```java
class NoFieldModel {}

class AllIgnoredModel {
    @ExcelIgnore
    private int number;
    
    @ExcelIgnore
    private Character character;
}
```

### writer:

```java
Javaxcel.newInstance()
        .writer(workbook, NoFieldModel.class); // Occurs exception.
Javaxcel.newInstance()
        .writer(workbook, AllIgnoredModel.class); // Occurs exception.
```

If you try to write with the class that has no targeted fields, `ExcelWriter` will throw exception.

<br>

### reader:

```java
List<NoFieldModel> noFieldModels = Javaxcel.newInstance()
        .reader(workbook, NoFieldModel.class); // Occurs exception.
List<AllIgnoredModel> allIgnoredModels = Javaxcel.newInstance()
        .reader(workbook, AllIgnoredModel.class); // Occurs exception.
```

If you try to write with the class that has no targeted fields, `ExcelReader` will throw exception.

<br><br>

## Model that extends class

```java
class EducationalProduct extends Product {
    private int[] targetAges;
    private String goals;
    private Product related; // Unknown type.
    private LocalDate date;
    private LocalTime time;
    private LocalDateTime dateTime;
}

// ...

EducationalProduct eduProduct = EducationalProduct.builder()
        .serialNumber(10001)
        .name("Mathematics puzzle toys for kids")
        .accessId("1a57-4055-a75b-98e4")
        .width(18.0)
        .depth(6)
        .height(20)
        .weight(340.07)
        .targetAges(4,5,6,7,8,9)
        .goals("Develop intelligence")
        .related(new Product())
        .date(LocalDate.now())
        .time(LocalTime.now())
        .dateTime(LocalDateTime.now())
        .build();
List<EducationalProduct> eduProducts = Collections.singletonList(eduProduct);
```

There is a list that contains a `EducationalProduct`.

<br>

### writer:

```java
Javaxcel.newInstance()
        .writer(workbook, EducationalProduct.class)
        .write(out, list);
```

| targetAges         | goals                | related                              | date       | time     | dateTime            |
| ------------------ | -------------------- | ------------------------------------ | ---------- | -------- | ------------------- |
| [4, 5, 6, 7, 8, 9] | Develop intelligence | com.github.javaxcel.Product@4bf558aa | 2020-09-13 | 11:54:26 | 2020-09-13 11:54:26 |

It writes the declared own fields, not including the inherited fields.

That's default.

<br>

```java
@ExcelModel(includeSuper = true)
class EducationalProduct extends Product { /* ... */ }
```

If you annotate `@ExcelModel` and assign true into `includeSuper()`, it writes including the inherited fields.

The result is

| serialNumber | name                             | accessId            | width | depth | height | weight | targetAges         | goals                | related                             | date       | time     | dateTime            |
| ------------ | -------------------------------- | ------------------- | ----- | ----- | ------ | ------ | ------------------ | -------------------- | ----------------------------------- | ---------- | -------- | ------------------- |
| 10001        | Mathematics puzzle toys for kids | 1a57-4055-a75b-98e4 | 18.0  | 6.0   | 20.0   | 340.07 | [4, 5, 6, 7, 8, 9] | Develop intelligence | com.github.javaxcel.Product@b1a58a3 | 2020-09-13 | 11:54:26 | 2020-09-13 11:54:26 |


CHANGELOG.md
# Table of Contents

- [v0.9.0](#v090):
- [v0.8.2](#v082): 2022-07-10
- [v0.8.1](#v081): 2022-04-19
- [v0.8.0](#v080): 2022-04-15
- [v0.7.5](#v075): 2021-11-06
- [v0.7.4](#v074): 2021-05-27
- [v0.7.3](#v073): 2021-05-27
- [v0.7.2](#v072): 2021-05-25
- [v0.7.1](#v071): 2021-05-25
- [v0.7.0](#v070): 2021-05-12
- [v0.6.0](#v060): 2021-04-11
- [v0.5.3](#v053): 2020-12-06
- [v0.5.2](#v052): 2020-12-03
- [v0.5.1](#v051): 2020-11-09
- [v0.5.0](#v050): 2020-11-06
- [v0.4.2](#v042): 2020-10-17
- [v0.4.1](#v041): 2020-10-11
- [v0.4.0](#v040): 2020-09-27
- [v0.3.1](#v031): 2020-09-13
- [v0.3.0](#v030): 2020-09-12
- [v0.2.1](#v021): 2020-09-07
- [v0.1.0](#v010): 2020-08-30

# v0.9.0

### Modification

- 🔥 Remove: useless packages `constant`, `model` — `ConversionType`, `ConverterType`, `Column`
- 🔥 Remove: unused exceptions `GettingFieldValueException`, `NoTargetedConstructorException`
  , `NotExistConverterException`
  , `SettingFieldValueException`
- ♻️ Refactor: package `converter`
- ♻️ Refactor: handlers for java time
- ♻️ Replace: default constructor of strategy `Filter` with `Filter(boolean)`
- 🚚 Move: package `util.resolver` to `in.resolver`
- 🚚 Move: packages of implementation `ExcelTypeHandler`
- 🚚 Rename: implementations of converter
- ⚡️ Improve: support for handling field `Iterable` by converter
- 🔨 Modify: utility `FieldUtils.resolveActualType(Field)`
- 🔨 Modify: default policy of sheet name that increases suffix at 0 is changed to increase at 1.
- 🔨 Modify: type of parameter `convert(Map<String, Object>, Field)` in `ExcelReadConverter` is changed
  to `convert(Map<String, String>, Field)`
- 🔨 Modify:  generic return type of `Javaxcel.reader(Workbook)` is changed from `ExcelReader<Map<String, Object>>`
  to `ExcelReader<Map<String, String>>`
- 🔨 Modify: method `newInstance(ExcelTypeHandlerRegistry)` in `Javaxcel` is changed not to
  use `DefaultExcelTypeHandlerRegistry`.

### New features

- 📦️ Add: package `analysis`, `util.processor`
- ✨ Add: method `add(ExcelTypeHandler)` in `ExcelTypeHandlerRegistry`
- ✨ Add: strategies `UseGetters`, `UseSetters`
- ✨ Add: methods `resolveGetter(Field)`, `resolveSetter(Field)` in `FieldUtils`
- ✨ Add: type handlers for `Year`, `YearMonth`, `Month`, `MonthDay`, `Instant`

### Troubleshooting

- 🐞 Fix:

### Dependencies

- ➖ Remove: dependency `jsr305`
- ➕ Add: intransitive dependency `annotations`
- ⬆️ Upgrade: provided dependency `poi-ooxml` from `5.2.2` to `5.2.3`
- ⬆️ Upgrade: dependency `common-utils` from `0.9.0` to `0.13.0`
- ⬆️ Upgrade: test dependency `junit5` from `5.8.2` to `5.9.1`
- ⬆️ Upgrade: test dependency `assertj-core` from `3.22.0` to `3.23.1`
- ⬆️ Upgrade: test dependency `spock-core` from `2.1-groovy-3.0` to `2.3-groovy-4.0`
- ⬆️ Upgrade: test dependency `byte-buddy` from `1.12.9` to `1.12.18`
- ⬆️ Upgrade: test dependency `excel-streaming-reader` from `3.6.1` to `4.0.4`
- ⬆️ Upgrade: build dependency `gmavenplus-plugin` from `1.13.1` to `2.1.0`
- ⬆️ Upgrade: build dependency `maven-jar-plugin` from `3.2.0` to `3.3.0`
- ⬆️ Upgrade: build dependency `maven-javadoc-plugin` from `3.2.0` to `3.4.1`
- ⬆️ Upgrade: build dependency `maven-shade-plugin` from `3.3.0` to `3.4.1`

# v0.8.2

### Modification

- 🔥 Remove: useless type variable in `ExcelWriteConverter`
- 🔥 Remove: method `toMap(Object)` in `FieldUtils`
- 🚚 Move: package of classes from `com.github.javaxcel.out.strategy.ExcelWriteStrategy.*`
  to `com.github.javaxcel.out.strategy.impl.*`
- 🚚 Move: package of classes from `com.github.javaxcel.in.strategy.ExcelWriteStrategy.*`
  to `com.github.javaxcel.in.strategy.impl.*`
- ♻️ Increase: access privileges to `AbstractExcelWriteStrategy`, `AbstractExcelReadStrategy`
- ⚡️ Modify: `com.github.javaxcel.out.strategy.impl.Filter` to freeze header

### Troubleshooting

- 🐞 Fix: wrong import (`java.util.logging.Filter` => `com.github.javaxcel.out.strategy.impl.Filter`) in `ModelWriter`
  , `MapWriter`
- 🐞 Fix: compile error on jdk 9+ by using internal package `sun.util.locale.*` in `LocaleTypeHandler`

### Dependencies

- ⬆️ Upgrade: dependency `common-utils` from `0.8.0` to `0.9.0`

# v0.8.1

### Dependencies

- ⬆️ Upgrade: dependency `javaxcel-styler` from `0.2.1` to `0.2.2` --- nested dependency `poi-ooxml` provided
- ⬆️ Upgrade: test dependency `lombok` from `1.18.22` to `1.18.24`

# v0.8.0

### Modification

- 🚚 Move: package of classes from `com.github.javaxcel.out` to `com.github.javaxcel.out.core`
- 🚚 Move: subclasses `ExcelWriter` from `com.github.javaxcel.out` to `com.github.javaxcel.out.core.impl`
- 🚚 Rename: method `setDefaultValue(String)` to `setAllDefaultValues(String)` in `ExcelWriteConverterSupport`
- 🚚 Rename: annotation `@ExcelWriterExpression` to `@ExcelWriteExpression`
- 🚚 Rename: annotation `@ExcelReaderExpression` to `@ExcelReadExpression`
- ♻️ Refactor: implementations of `ExcelWriter`
- ♻️ Refactor: input/output conversion
- ♻️ Refactor: input/output core module
- ♻️ Replace: `ExcelWriterFactory`, `ExcelReaderFactory` with `Javaxcel`
- ⚡️ Resolve: generic type of `MapWriter`
- ⚡️ Support: non-default constructor, static factory method to instantiate model
- 🔥 Remove: exception `NoTargetedConstructorException`
- 🔥 Remove: methods `getSheetRange(Workbook)`, `getNumOfRows(File)` in `ExcelUtils`

### New features

- ✨ Add: entrypoint class `Javaxcel`
- ✨ Add: domain model `Column`
- ✨ Add: annotations `@ExcelModelCreator`, `@ExcelModelCreator.FieldName`
- ✨ Add: modules `AbstractExcelModelExecutableResolver`, `ExcelModelConstructorResolver`, `ExcelModelMethodResolver`
- ✨ Add: modules `ExcelTypeHandler`, `ExcelTypeHandlerRegistry`
- ✨ Add: methods `toMap(T)`, `toHeaderName(Field)`, `resolveFirst(Class, Object...)`, `resolveLast(Class, Object...)`
  in `FieldUtils`
- 🔊 Add: fluent exception messages

### Troubleshooting

- 🐞 Fix: wrong computation of targeted fields
- 🐞 Fix: security problem for changing value of final field

### Dependencies

- ♻️ Make: dependency `poi-ooxml` provided

src/main/java/com/github/javaxcel/out/MapWriter.java
/*
 * Copyright 2020 Javaxcel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.javaxcel.out;

import com.github.javaxcel.styler.ExcelStyleConfig;
import io.github.imsejin.common.assertion.Asserts;
import org.apache.poi.ss.usermodel.*;

import javax.annotation.Nullable;
import java.io.OutputStream;
import java.util.*;
import java.util.function.Predicate;

import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

/**
 * Excel writer for {@link Map}.
 *
 * @param <T> {@link Map}
 */
public class MapWriter<T extends Map<String, ?>> extends AbstractExcelWriter<T> {

    /**
     * @see #beforeWrite(OutputStream, List)
     */
    private final List<String> keys = new ArrayList<>();

    /**
     * Map of which key is the key and value is index number.
     *
     * @see #headerNames(List, List)
     * @see #beforeWrite(OutputStream, List)
     */
    private Map<String, Integer> indexedMap;

    private String defaultValue;

    /**
     * @param workbook excel workbook
     * @see com.github.javaxcel.factory.ExcelWriterFactory#create(Workbook)
     */
    public MapWriter(Workbook workbook) {
        super(workbook);
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> defaultValue(String defaultValue) {
        super.defaultValue(defaultValue);
        this.defaultValue = defaultValue;
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> sheetName(String sheetName) {
        super.sheetName(sheetName);
        return this;
    }

    /**
     * Rearranges the keys of {@link Map} with custom order.
     *
     * <p> If you export list of {@link Map} as a excel file,
     * column order is not guaranteed, unless the type of its instance is {@link LinkedHashMap}.
     * For example, the following list will be exported.
     *
     * <pre><code>
     * [
     *     {
     *         "serialNumber": 10000,
     *         "name": "Choco cereal",
     *         "apiId": "2a60-4973-aec0-685e",
     *         "width": null,
     *         "depth": 0.0,
     *         "height": 20.5,
     *         "weight": 580.5
     *     },
     *     {
     *         "serialNumber": 10001,
     *         "name": "Oatmeal cereal",
     *         "apiId": "f15d-384d-0a4b-97ec",
     *         "width": 10.2,
     *         "depth": 4.0,
     *         "height": 6.0,
     *         "weight": 575.0
     *     }
     * ]
     * </code></pre>
     *
     * <p> To rearrange the column order, place the keys in the order you want like this.
     *
     * <pre>{@code
     *     List<String> orderedKeys = Arrays.asList(
     *             "width" "depth", "height", "weight", "serialNumber", "name", "apiId");
     *
     *     ExcelWriterFactory.create(new SXSSFWorkbook())
     *             .headerNames(orderedKeys)
     *             .write(new FileOutputStream(file), list);
     * }</pre>
     *
     * <p> Then the columns will be arranged in the order you want.
     *
     * <pre><code>
     * +-------+-------+--------+--------+--------------+----------------+---------------------+
     * | width | depth | height | weight | serialNumber | name           | apiId               |
     * +-------+-------+--------+--------+--------------+----------------+---------------------+
     * |       | 0.0   | 20.5   | 580.5  | 10000        | Choco cereal   | 2a60-4973-aec0-685e |
     * +-------+-------+--------+--------+--------------+----------------+---------------------+
     * | 10.2  | 4.0   | 6.0    | 575.0  | 10001        | Oatmeal cereal | f15d-384d-0a4b-97ec |
     * +-------+-------+--------+--------+--------------+----------------+---------------------+
     * </code></pre>
     *
     * @param orderedKeys keys ordered as you want
     * @return {@link MapWriter}
     * @throws IllegalArgumentException if ordered keys is null or empty
     */
    @Override
    public MapWriter<T> headerNames(List<String> orderedKeys) {
        return headerNames(orderedKeys, null);
    }

    /**
     * Rearranges the keys of {@link Map} with custom order and sets header names.
     *
     * <p> If you export list of {@link Map} as a excel file,
     * column order is not guaranteed, unless the type of its instance is {@link LinkedHashMap}.
     * For example, the following list will be exported.
     *
     * <pre><code>
     * [
     *     {
     *         "serialNumber": 10000,
     *         "name": "Choco cereal",
     *         "apiId": "2a60-4973-aec0-685e",
     *         "width": null,
     *         "depth": 0.0,
     *         "height": 20.5,
     *         "weight": 580.5
     *     },
     *     {
     *         "serialNumber": 10001,
     *         "name": "Oatmeal cereal",
     *         "apiId": "f15d-384d-0a4b-97ec",
     *         "width": 10.2,
     *         "depth": 4.0,
     *         "height": 6.0,
     *         "weight": 575.0
     *     }
     * ]
     * </code></pre>
     *
     * <p> To rearrange the column order, place the keys in the order you want
     * to the first argument. To change the header names, place the names
     * you want them changed to in the custom order to the second argument like this.
     *
     * <pre>{@code
     *     List<String> orderedKeys = Arrays.asList(
     *             "width" "depth", "height", "weight", "serialNumber", "name", "apiId");
     *     List<String> name = Arrays.asList(
     *             "WIDTH" "DEPTH", "HEIGHT", "WEIGHT", "SERIAL_NUMBER", "NAME", "API_ID");
     *
     *     ExcelWriterFactory.create(new SXSSFWorkbook())
     *             .headerNames(orderedKeys, name)
     *             .write(new FileOutputStream(file), list);
     * }</pre>
     *
     * <p> Then the column order and the names will be changed you want.
     *
     * <pre><code>
     * +-------+-------+--------+--------+---------------+----------------+---------------------+
     * | WIDTH | DEPTH | HEIGHT | WEIGHT | SERIAL_NUMBER | NAME           | API_ID              |
     * +-------+-------+--------+--------+---------------+----------------+---------------------+
     * |       | 0.0   | 20.5   | 580.5  | 10000         | Choco cereal   | 2a60-4973-aec0-685e |
     * +-------+-------+--------+--------+---------------+----------------+---------------------+
     * | 10.2  | 4.0   | 6.0    | 575.0  | 10001         | Oatmeal cereal | f15d-384d-0a4b-97ec |
     * +-------+-------+--------+--------+---------------+----------------+---------------------+
     * </code></pre>
     *
     * @param orderedKeys keys ordered as you want
     * @param headerNames header names in key order
     * @return {@link MapWriter}
     * @throws IllegalArgumentException if ordered keys is null or empty
     * @throws IllegalArgumentException if num of ordered keys is not equal to num of header names
     */
    public MapWriter<T> headerNames(List<String> orderedKeys, @Nullable List<String> headerNames) {
        Asserts.that(orderedKeys)
                .as("Ordered keys is now allowed to be null or empty")
                .hasElement();

        if (headerNames != null) {
            // Validates ordered keys and header names.
            Asserts.that(orderedKeys)
                    .as("The number of ordered keys is not equal to the number of header names")
                    .isSameSize(headerNames);

            super.headerNames(headerNames);
        }

        // Creates indexed map for rearrange.
        Map<String, Integer> indexedMap = new HashMap<>();
        for (int i = 0; i < orderedKeys.size(); i++) {
            indexedMap.put(orderedKeys.get(i), i);
        }

        this.indexedMap = indexedMap;

        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> unrotate() {
        super.unrotate();
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> filter() {
        super.filter();
        return this;
    }

    ///////////////////////////////////// Decoration //////////////////////////////////////

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> headerStyle(ExcelStyleConfig config) {
        return headerStyles(config);
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> headerStyles(ExcelStyleConfig... configs) {
        super.headerStyles(configs);
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> bodyStyle(ExcelStyleConfig config) {
        return bodyStyles(config);
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> bodyStyles(ExcelStyleConfig... configs) {
        super.bodyStyles(configs);
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> autoResizeColumns() {
        super.autoResizeColumns();
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> hideExtraRows() {
        super.hideExtraRows();
        return this;
    }

    /**
     * {@inheritDoc}
     *
     * @return {@link MapWriter}
     */
    @Override
    public MapWriter<T> hideExtraColumns() {
        super.hideExtraColumns();
        return this;
    }

    //////////////////////////////////////// Hooks ////////////////////////////////////////

    /**
     * {@inheritDoc}
     *
     * <p> Unlike {@link ModelWriter}, {@link MapWriter} cannot validate header names,
     * header styles and body styles before {@link #write(OutputStream, List)} is invoked
     * because the only way to get {@link Map} exists in that method. So it does using this hook.
     */
    @Override
    protected void beforeWrite(OutputStream out, List<T> list) {
        // To write a header, this doesn't allow to accept empty list of maps.
        Asserts.that(list)
                .as("List of maps is not allowed to be empty")
                .hasElement();

        // Gets all the maps' keys.
        this.keys.addAll(list.stream().flatMap(it -> it.keySet().stream()).distinct().collect(toList()));

        if (this.indexedMap != null) {
            // Invalidates the number of ordered keys and their each element.
            Set<String> orderedKeys = this.indexedMap.keySet();

            Asserts.that(this.keys)
                    .as("Ordered keys are at variance with maps' keys\nmaps' keys: {0}\nordered keys: {1}",
                            this.keys, orderedKeys)
                    .isSameSize(orderedKeys)
                    .containsAll(orderedKeys);

            // Rearranges the keys as you want: it changes order of columns.
            this.keys.sort(comparing(this.indexedMap::get));
        }

        final int numOfKeys = this.keys.size();
        Predicate<CellStyle[]> validator = them -> them.length == 1 || them.length == numOfKeys;

        // Validates the number of header styles.
        if (this.headerStyles != null) {
            Asserts.that(validator.test(this.headerStyles))
                    .as("Number of header styles({0}) must be 1 or equal to number of maps' keys({1})",
                            this.headerStyles.length, numOfKeys)
                    .isTrue();
        }

        // Validates the number of body styles.
        if (this.bodyStyles != null) {
            Asserts.that(validator.test(this.bodyStyles))
                    .as("Number of body styles({0}) must be 1 or equal to number of maps' keys({1})",
                            this.bodyStyles.length, numOfKeys)
                    .isTrue();
        }
    }

    /**
     * {@inheritDoc}
     *
     * <p> If the header names are not set through {@link #headerNames(List, List)},
     * this method brings the values from {@link Map#keySet()}.
     *
     * @see Map#keySet()
     */
    @Override
    protected void ifHeaderNamesAreEmpty(List<String> headerNames) {
        headerNames.addAll(this.keys);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void writeToSheet(Sheet sheet, List<T> list) {
        final int numOfMaps = list.size();
        final int numOfKeys = this.keys.size();

        for (int i = 0; i < numOfMaps; i++) {
            T map = list.get(i);

            // Skips the first row that is header.
            Row row = sheet.createRow(i + 1);

            for (int j = 0; j < numOfKeys; j++) {
                Object value = map.get(this.keys.get(j));
                Cell cell = row.createCell(j);

                // Not allows empty string to be written.
                if (value != null && !"".equals(value)) {
                    cell.setCellValue(value.toString());
                } else if (this.defaultValue != null) {
                    cell.setCellValue(this.defaultValue);
                }

                if (this.bodyStyles == null) continue;

                // Sets styles to body's cell.
                CellStyle bodyStyle = this.bodyStyles.length == 1
                        ? this.bodyStyles[0] : this.bodyStyles[j];

                // There is possibility that 'bodyStyles' has null elements, if you set 'NoStyleConfig'.
                if (bodyStyle != null) cell.setCellStyle(bodyStyle);
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected int getNumOfColumns() {
        return this.keys.size();
    }

}
